File: src/main/java/de/esymetric/jerusalem/utils/FileBasedHashMap.java
package de.esymetric.jerusalem.utils;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.File;
import java.io.IOException;


public class FileBasedHashMap {

	final static int HASH_ARRAY_SIZE = 0x100000;
	final static int HASH_ARRAY_MASK = HASH_ARRAY_SIZE - 1;
	final static int SENTENCE_LENGTH = 12; // 12 bytes: int key, int value, int
	// nextIndex

	BufferedRandomAccessFile raf;
	BufferedRandomAccessFileCache rafCache = new BufferedRandomAccessFileCache();

	boolean isEmpty = false;
	long fileLength;
	
	int countGets, countGetReads;
	
	public float getAvgGetAccessNumberOfReads() { 
		float r = (float)countGetReads / (float)countGets;
		countGetReads = 0; countGets = 0;
		return r;
	}
	
	public int getAndClearNumberOfFileChanges() {
		return rafCache.getAndClearNumberOfFileChanges();
	}

	public void open(String filePath, boolean readOnly) {
		isEmpty = false;
		fileLength = 0L;
		
		if (readOnly && !new File(filePath).exists()) {
			isEmpty = true;
			return;
		}

		try {
			raf = rafCache.getRandomAccessFile(filePath, readOnly);
			/*
			if( raf != null ) raf.close();
			raf = new RandomAccessFile(filePath, readOnly ? "r" : "rw");
			*/
			if (raf.length() < (long) SENTENCE_LENGTH * (long) HASH_ARRAY_SIZE) {
				byte[] buf = new byte[1024];
				for (int i = 0; i < HASH_ARRAY_SIZE / 1024 * SENTENCE_LENGTH; i++)
					raf.write(buf);
				System.out.print("$");
			}
			
			fileLength = raf.length();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public void close() {
		rafCache.close();
	}
	
	public void put(int key, int value) {

		try {
			int arrayIndex = key & HASH_ARRAY_MASK;
			long pos = (long) arrayIndex * (long) SENTENCE_LENGTH;
			raf.seek(pos);
			int readKey = raf.readInt();
			raf.readInt();
			int readNextIndex = raf.readInt();
			
			if (readKey == 0) {
				raf.seek(pos);
				raf.writeInt(key);
				raf.writeInt(value);
			} else {
				
				int newIndex = (int) (fileLength / (long) SENTENCE_LENGTH);
				raf.seek(fileLength);
				raf.writeInt(key);
				raf.writeInt(value);
				raf.writeInt(readNextIndex); 
				fileLength += SENTENCE_LENGTH;

				raf.seek(pos + 8L);
				raf.writeInt(newIndex);
				
				
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public int get(int key) {
		if (isEmpty)
			return -1;

		try {
			int arrayIndex = key & HASH_ARRAY_MASK;
			long pos = (long) arrayIndex * (long) SENTENCE_LENGTH;
			countGets++;
			countGetReads++;
			byte[] buf = new byte[12];
			for (;;countGetReads++) {
				raf.seek(pos);
				raf.read(buf);
				
				ByteArrayInputStream bais = new ByteArrayInputStream(buf);
				DataInputStream dis = new DataInputStream(bais);
				int foundKey = dis.readInt();
				int foundValue = dis.readInt();
				int nextIndex = dis.readInt();
				dis.close();
				bais.close();
				if (foundKey == key)
					return foundValue;
				if (nextIndex <= 0)
					return -1;
				else {
					pos = (long) nextIndex * (long) SENTENCE_LENGTH;
				}
			}
		} catch (IOException e) {
			e.printStackTrace();
			return -1;
		}
	}

}

File: src/main/java/de/esymetric/jerusalem/utils/FileBasedHashMap.kt
package de.esymetric.jerusalem.utils

import java.io.ByteArrayInputStream
import java.io.DataInputStream
import java.io.File
import java.io.IOException

class FileBasedHashMap {
    // nextIndex
    var raf: BufferedRandomAccessFile? = null
    var rafCache = BufferedRandomAccessFileCache()
    var isEmpty = false
    var fileLength: Long = 0
    var countGets = 0
    var countGetReads = 0
    val avgGetAccessNumberOfReads: Float
        get() {
            val r = countGetReads.toFloat() / countGets.toFloat()
            countGetReads = 0
            countGets = 0
            return r
        }
    val andClearNumberOfFileChanges: Int
        get() = rafCache.andClearNumberOfFileChanges

    fun open(filePath: String?, readOnly: Boolean) {
        isEmpty = false
        fileLength = 0L
        if (readOnly && !File(filePath).exists()) {
            isEmpty = true
            return
        }
        try {
            raf = rafCache.getRandomAccessFile(filePath, readOnly)
            if (raf!!.length() < SENTENCE_LENGTH.toLong() * HASH_ARRAY_SIZE.toLong()) {
                val buf = ByteArray(1024)
                for (i in 0 until HASH_ARRAY_SIZE / 1024 * SENTENCE_LENGTH) raf?.write(buf)
                print("$")
            }
            fileLength = raf!!.length()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun close() {
        rafCache.close()
    }

    fun put(key: Int, value: Int) {
        try {
            val arrayIndex = key and HASH_ARRAY_MASK
            val pos = arrayIndex.toLong() * SENTENCE_LENGTH.toLong()
            raf!!.seek(pos)
            val readKey = raf!!.readInt()
            raf!!.readInt()
            val readNextIndex = raf!!.readInt()
            if (readKey == 0) {
                raf!!.seek(pos)
                raf!!.writeInt(key)
                raf!!.writeInt(value)
            } else {
                val newIndex = (fileLength / SENTENCE_LENGTH.toLong()).toInt()
                raf!!.seek(fileLength)
                raf!!.writeInt(key)
                raf!!.writeInt(value)
                raf!!.writeInt(readNextIndex)
                fileLength += SENTENCE_LENGTH.toLong()
                raf!!.seek(pos + 8L)
                raf!!.writeInt(newIndex)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    operator fun get(key: Int): Int {
        if (isEmpty) return -1
        try {
            val arrayIndex = key and HASH_ARRAY_MASK
            var pos = arrayIndex.toLong() * SENTENCE_LENGTH.toLong()
            countGets++
            countGetReads++
            val buf = ByteArray(12)
            while (true) {
                raf!!.seek(pos)
                raf!!.read(buf)
                val bais = ByteArrayInputStream(buf)
                val dis = DataInputStream(bais)
                val foundKey = dis.readInt()
                val foundValue = dis.readInt()
                val nextIndex = dis.readInt()
                dis.close()
                bais.close()
                if (foundKey == key) return foundValue
                pos = if (nextIndex <= 0) return -1 else {
                    nextIndex.toLong() * SENTENCE_LENGTH.toLong()
                }
                countGetReads++
            }
        } catch (e: IOException) {
            e.printStackTrace()
            return -1
        }
    }

    companion object {
        const val HASH_ARRAY_SIZE = 0x100000
        const val HASH_ARRAY_MASK = HASH_ARRAY_SIZE - 1
        const val SENTENCE_LENGTH = 12 // 12 bytes: int key, int value, int
    }
}